### PR TITLE
Simplify cluster configuration

### PR DESCRIPTION
**Pull request by kuujo**
This PR simplifies the cluster configuration format. Rather than requiring two separate sections for core/bootstrap nodes, the `CORE` nodes are just extracted from a single list of nodes for the cluster. To join a consensus cluster, a static set of `CORE` nodes must be defined. To join a primary-backup-based cluster, a single active peer must be defined or multicast bootstrap must be enabled.

### CODE REVIEW

File: core/src/main/java/io/atomix/core/Atomix.java [hunk 1]
```diff
@@ -472,23 +471,38 @@ private static ManagedMessagingService buildMessagingService(AtomixConfig config
...
   /**
    * Builds a bootstrap metadata service.
    */
   private static ManagedBootstrapMetadataService buildBootstrapMetadataService(AtomixConfig config) {
-    return new DefaultBootstrapMetadataService(config.getClusterConfig().getBootstrapConfig());
+    boolean hasCoreNodes = config.getClusterConfig().getNodes().stream().anyMatch(node -> node.getType() == Node.Type.CORE);
+    ClusterMetadata metadata = ClusterMetadata.builder()
+        .withNodes(config.getClusterConfig().getNodes()
+            .stream()
+            .filter(node -> !hasCoreNodes || node.getType() == Node.Type.CORE)
```
Comment (johnou) [inline core/src/main/java/io/atomix/core/Atomix.java]: Do we really want clients in here too if no core nodes are defined?

Comment (kuujo) [reply core/src/main/java/io/atomix/core/Atomix.java]: Hmm... probably not

Comment (johnou) [reply core/src/main/java/io/atomix/core/Atomix.java]: the way I read it at least is that if no core nodes are defined client nodes will be considered bootstrap nodes.

Comment (kuujo) [reply core/src/main/java/io/atomix/core/Atomix.java]: That is indeed true

Comment (kuujo) [reply core/src/main/java/io/atomix/core/Atomix.java]: Fixed

**Comment by johnou** [comment]
@kuujo gitignore core/data/?

File: core/src/test/java/io/atomix/core/AbstractAtomixTest.java [hunk 2]
```diff
@@ -50,36 +51,35 @@ public static void setupAtomix() throws Exception {
...
-    Collection<Node> bootstrapNodes = bootstrapIds.stream()
-        .map(nodeId -> Node.builder(String.valueOf(nodeId))
-            .withType(Node.Type.CORE)
-            .withAddress("localhost", BASE_PORT + nodeId)
-            .build())
+    Collection<Node> nodes = Stream.concat(
+        nodeIds.stream()
+            .map(nodeId -> Node.builder(String.valueOf(nodeId))
+                .withType(Node.Type.CORE)
+                .withAddress("localhost", BASE_PORT + nodeId)
```
Comment (johnou) [inline core/src/test/java/io/atomix/core/AbstractAtomixTest.java]: find available port helper method?

**Comment by johnou** [comment]
one more nit and lgtm.

File: core/src/test/java/io/atomix/core/AbstractAtomixTest.java (hunk 2 above)
Comment (kuujo) [reply core/src/test/java/io/atomix/core/AbstractAtomixTest.java]: Added that in #482

**Pull request merged**